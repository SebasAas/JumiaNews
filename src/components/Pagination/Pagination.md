## Importing Component

```js static
import Pagination from '<RootFolder>';
```

<hr/>

## Basic Pagination:

```html
<Pagination />
```

<hr/>

## Pagination in exact page

### Expected
- String or Object

### Default
- 1

### Return
- Pagination Component start on page 5

```html
<Pagination page='5' />
```

<hr/>

## Pagination with pages remain:

### Expected

- Number 

### Return
- 5 (By default pagination do the math by 10 to get the remaining pages)

```html
<Card hits='50' />
```

<hr/>

## Pagination with handle change page:

### Expected

- Function (setState) used to set the new value of page on click button prev and next


```html
<Card setPage={fn()} />
```
