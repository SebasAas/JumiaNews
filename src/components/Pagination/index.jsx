import React, { useEffect, useState } from 'react'

// Utils
import scrollToTop from '../../utils/scrollToTop.js';

// Styles
import styled from 'styled-components';
import { primaryOrange } from '../../theme/colors.js'

// Icons
import { ArrowLeft, ArrowRight } from '../../assets/Icons';

function Pagination({ page = 1, setPage, hits }) {
  const [pages, setPages] = useState(page);

  useEffect(() => {
    scrollToTop()
  }, [])

  useEffect(() => {
    if (page !== pages) {
      setPages(page)
    }
  }, [page])

  return (
    <PaginationContainer>
      <PrevButton data-testid="prev-button" disabled={pages <= 1} onClick={() => { setPages(prev => prev - 1); setPage(prev => prev - 1); scrollToTop() }}>
        <ArrowLeft />
      </PrevButton>
      <NumberPage><span>{pages}</span> of <span>{hits ? Math.floor(hits / 10) : ''}</span></NumberPage>
      <NextButton data-testid="next-button" disabled={pages * 10 === hits} onClick={() => { setPages(prev => prev + 1); setPage(prev => prev + 1); scrollToTop() }}>
        <ArrowRight />
      </NextButton>
    </PaginationContainer>
  )
}

const PaginationContainer = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
  min-width: 250px;
  margin-top: 30px;
`

const PrevButton = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  background: ${primaryOrange};
  color: white;
  border-radius: 50%;
  padding: 7px;
  cursor: pointer;

  &:disabled {
    color: #d6d6d6;
  }
`

const NextButton = styled.button`
  display: flex;
  justify-content: center;
  align-items: center;
  background: ${primaryOrange};
  color: white;
  border-radius: 50%;
  padding: 7px;
  cursor: pointer;

  &:disabled {
    color: #bebebe;
  }
`

const NumberPage = styled.p`
    font-weight: 500;
    font-size: 18px;
    letter-spacing: 0.05em;
`

export default Pagination
