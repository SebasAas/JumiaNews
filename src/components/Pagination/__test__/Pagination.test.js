import React from 'react';
import "@testing-library/jest-dom/extend-expect";
import { render, fireEvent, waitFor } from "@testing-library/react";
import userEvent from '@testing-library/user-event'
// import { MemoryRouter } from 'react-router-dom'

// Components
import Pagination from '../index'


describe('Rendering Filter Component', () => {

  afterEach(() => {
    jest.resetAllMocks();
  });
  afterAll(() => {
    jest.clearAllMocks();
  });
  beforeEach(() => {
    window.scrollTo = jest.fn();
  });

  test('should start in the first page', () => {
    const {
      getByText,
      // debug
    } = render(<Pagination />)

    expect(getByText('1')).toBeVisible()
  })

  test('prev button should be disabled at first page', () => {
    const { getByTestId } = render(<Pagination />)

    expect(getByTestId('prev-button')).toBeDisabled();
  })

  test('next button should be disabled at last page', () => {
    // Are showed 10 news per page
    const { getByTestId } = render(<Pagination page={5} hits={50} />)

    expect(getByTestId('next-button')).toBeDisabled();
  })

  test('prev and next button should be enabled between first and end page', () => {

    const { getByTestId } = render(<Pagination page={2} hits={50} />)

    expect(getByTestId('prev-button')).not.toBeDisabled();
    expect(getByTestId('next-button')).not.toBeDisabled();
  })

  test('should add 1 page on click next button', async () => {
    const setPage = jest.fn();

    const { getByTestId, getByText } = render(<Pagination page={1} setPage={setPage} />)

    expect(getByText('1')).toBeVisible()

    userEvent.click(getByTestId('next-button'))

    // Validations
    expect(setPage).toBeCalledTimes(1)

    expect(getByText('2')).toBeVisible()
  })

  test('should go back 2 pages on click prev button', async () => {
    const setPage = jest.fn();

    const { getByTestId, getByText } = render(<Pagination page={4} setPage={setPage} />)
    expect(getByText('4')).toBeVisible()

    fireEvent.click(getByTestId('prev-button'))
    fireEvent.click(getByTestId('prev-button'))

    expect(setPage).toBeCalledTimes(2)

    expect(getByText('2')).toBeVisible()
  })

  test('should show remaining pages', () => {
    const hits = 50;

    const { getByText } = render(<Pagination page={1} hits={hits} />)

    // As we show 10 news per page, 50/10 = 5 remeaning pages
    expect(getByText(Math.floor(hits / 10))).toBeVisible()
  })

  test('should match snapshot pagination', () => {
    const hits = 50;
    const { container } = render(<Pagination page={1} hits={hits} />);
    expect(container).toMatchSnapshot()
  })

})