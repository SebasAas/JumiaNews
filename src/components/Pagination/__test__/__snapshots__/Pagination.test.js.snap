// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Rendering Filter Component should match snapshot pagination 1`] = `
<div>
  <div
    class="sc-bdvvtL bMDOoU"
  >
    <button
      class="sc-gsDKAQ dAOncb"
      data-testid="prev-button"
      disabled=""
    >
      <svg
        height="24px"
        version="1.1"
        viewBox="0 0 32 32"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <g
          id="surface1"
        >
          <path
            d="M 20.179688 8.070312 C 20.085938 8.101562 17.824219 10.007812 15.148438 12.300781 C 9.863281 16.832031 10 16.695312 10 17.335938 C 10 17.976562 9.863281 17.84375 14.957031 22.199219 C 17.539062 24.414062 19.761719 26.3125 19.898438 26.417969 C 20.226562 26.679688 20.6875 26.789062 21.039062 26.695312 C 21.625 26.539062 21.992188 26.085938 22.050781 25.457031 C 22.105469 24.792969 22.230469 24.917969 17.664062 21.019531 C 15.386719 19.074219 13.492188 17.449219 13.449219 17.40625 C 13.386719 17.351562 14.199219 16.617188 17.53125 13.773438 C 21.144531 10.6875 21.710938 10.175781 21.84375 9.914062 C 22.1875 9.230469 21.925781 8.476562 21.238281 8.117188 C 20.988281 7.988281 20.460938 7.96875 20.179688 8.070312 Z M 20.179688 8.070312 "
            style="stroke: none; fill-rule: nonzero; fill: currentColor; fill-opacity: 1;"
          />
        </g>
      </svg>
    </button>
    <p
      class="sc-hKwDye gozZfm"
    >
      <span>
        1
      </span>
       of 
      <span>
        5
      </span>
    </p>
    <button
      class="sc-dkPtRN lkCOSG"
      data-testid="next-button"
    >
      <svg
        height="24px"
        version="1.1"
        viewBox="0 0 32 32"
        width="24px"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <g
          id="surface1"
        >
          <path
            d="M 10.757812 6.789062 C 10.167969 7.082031 9.898438 7.707031 10.085938 8.34375 C 10.15625 8.574219 10.269531 8.738281 10.511719 8.976562 C 10.695312 9.148438 12.59375 10.789062 14.738281 12.617188 C 16.875 14.449219 18.625 15.960938 18.617188 15.988281 C 18.617188 16.011719 16.886719 17.507812 14.773438 19.3125 C 9.917969 23.457031 10.148438 23.242188 10.03125 23.644531 C 9.695312 24.773438 10.867188 25.761719 11.894531 25.210938 C 12.261719 25.011719 21.574219 17.023438 21.789062 16.726562 C 22.1875 16.164062 22.105469 15.5 21.582031 14.988281 C 21.417969 14.832031 19.195312 12.917969 16.644531 10.738281 C 13.167969 7.761719 11.949219 6.757812 11.761719 6.699219 C 11.386719 6.585938 11.117188 6.613281 10.757812 6.789062 Z M 10.757812 6.789062 "
            style="stroke: none; fill-rule: nonzero; fill: currentColor; fill-opacity: 1;"
          />
        </g>
      </svg>
    </button>
  </div>
</div>
`;
